accord-core/src/main/java/accord/messages/ReadData.java

```diff
@@ -222,48 +230,72 @@ protected final Action actionForStatus(SaveStatus status)
     }
 
     @Override
-    public synchronized void onChange(SafeCommandStore safeStore, SafeCommand safeCommand)
+    public synchronized boolean notify(SafeCommandStore safeStore, SafeCommand safeCommand)
     {
         Command command = safeCommand.current();
         logger.trace("{}: updating as listener in response to change on {} with status {} ({})",
                      this, command.txnId(), command.status(), command);
 
-        int id = safeStore.commandStore().id();
-        if (state != State.PENDING || !waitingOn.get(id))
-        {
-            removeListener(safeStore, safeCommand);
-            return;
-        }
+        int storeId = safeStore.commandStore().id();
+        if (state != State.PENDING)
+            return false;
 
         switch (actionForStatus(command.saveStatus()))
         {
             default: throw new AssertionError("Unhandled Action: " + actionForStatus(command.saveStatus()));
-
             case WAIT:
-                break;
+                return true;
 
             case OBSOLETE:
-                onFailure(safeStore, Redundant, null);
-                break;
+                onFailure(Redundant, null);
+                return false;
 
             case EXECUTE:
-                if (!reading.get(id))
+                if (!reading.get(storeId))
                 {
-                    reading.set(safeStore.commandStore().id());
+                    if (reading == waitingOn)
+                    {
+                        reading = new BitSet();
+                        reading.or(waitingOn);
+                    }
+                    if (!waitingOn.get(storeId))
+                    {
+                        waitingOn.set(storeId);
+                        ++waitingOnCount;
+                    }
+                    reading.set(storeId);
                     logger.trace("{}: executing read", command.txnId());
                     read(safeStore, command);
+                    if (notReady != null)
+                        notReady = notReady.without(executeRanges(safeStore));
                 }
+                return true;
         }
     }
 
+    Ranges executeRanges(SafeCommandStore safeStore)
+    {
+        // TODO (required): is this correct for all cases incl awaitsOnlyDeps etc
+        return safeStore.ranges().allAt(executeAtEpoch);
+    }
+
     @Override
     public synchronized void accept(CommitOrReadNack reply, Throwable failure)
     {
         // Unless failed always ack to indicate setup has completed otherwise the counter never gets to -1
         if ((reply == null || !reply.isFinal()) && failure == null)
-            onOneSuccess(null, null);
+        {
+            onOneSuccess(-1, null);
+            if (state == State.PENDING)
+            {
+                long replyTimeout = node.agent().replyTimeout(replyContext, MILLISECONDS);
+                timeout = node.requestTimeouts().register(this, replyTimeout, MILLISECONDS);
```

Comment:
This is timing out the local operation on the replica performing the read, so that it does not waste time actually performing the read (or retaining state for it in memory).

While the coordinator will have correctly timed out the read, the replica it contacted may have an unlimited number of reads waiting to be served for any number of transactions.